File: src/main/java/com/kobe/warehouse/service/report/StockRotationReportService.java
package com.kobe.warehouse.service.report;

import com.kobe.warehouse.domain.enumeration.CategorieABC;
import com.kobe.warehouse.service.dto.report.StockRotationDTO;
import java.util.List;
import java.util.Map;

public interface StockRotationReportService {
    /**
     * Get all stock rotation data
     *
     * @return List of stock rotation records
     */
    List<StockRotationDTO> getAllStockRotation();

    /**
     * Get stock rotation filtered by product category (famille)
     *
     * @param categorie The category name to filter by
     * @return List of stock rotation records for the category
     */
    List<StockRotationDTO> getStockRotationByCategory(String categorie);

    /**
     * Get stock rotation filtered by ABC classification
     *
     * @param categorieABC The ABC classification to filter by (A, B, C)
     * @return List of stock rotation records for the ABC classification
     */
    List<StockRotationDTO> getStockRotationByABCClassification(CategorieABC categorieABC);

    /**
     * Get count of products by ABC classification
     *
     * @return Map of ABC classification to count
     */
    Map<CategorieABC, Long> getStockRotationCountByABCClassification();

    /**
     * Get slow moving products (classification = C)
     *
     * @return List of slow moving products
     */
    List<StockRotationDTO> getSlowMovingProducts();

    /**
     * Get stock rotation data with pagination
     *
     * @param page Page number (0-indexed)
     * @param size Page size
     * @return List of stock rotation records for the page
     */
    List<StockRotationDTO> getStockRotationPaginated(int page, int size);

    /**
     * Get total count of stock rotation records
     *
     * @return Total count
     */
    long getStockRotationCount();

    /**
     * Get stock rotation by ABC classification with pagination
     *
     * @param categorieABC The ABC classification to filter by
     * @param page Page number (0-indexed)
     * @param size Page size
     * @return List of stock rotation records for the page
     */
    List<StockRotationDTO> getStockRotationByABCPaginated(CategorieABC categorieABC, int page, int size);

    /**
     * Get count of stock rotation records by ABC classification
     *
     * @param categorieABC The ABC classification to count
     * @return Count of records
     */
    long getStockRotationCountByABC(CategorieABC categorieABC);
}

File: pharmaSmart-app/src/main/webapp/app/entities/reports/services/stock-rotation-report.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpResponse, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { SERVER_API_URL } from 'app/app.constants';
import { IStockRotation, CategorieABC } from 'app/shared/model/report/stock-rotation.model';

type EntityArrayResponseType = HttpResponse<IStockRotation[]>;
type CountResponseType = HttpResponse<Record<CategorieABC, number>>;

@Injectable({ providedIn: 'root' })
export class StockRotationReportService {
  private readonly resourceUrl = SERVER_API_URL + 'api/stock/rotation';
  private readonly http = inject(HttpClient);

  /**
   * Get all stock rotation data
   */
  getAllStockRotation(): Observable<EntityArrayResponseType> {
    return this.http.get<IStockRotation[]>(this.resourceUrl, { observe: 'response' });
  }

  /**
   * Get stock rotation filtered by product category (famille)
   */
  getStockRotationByCategory(categorie: string): Observable<EntityArrayResponseType> {
    const params = new HttpParams().set('categorie', categorie);
    return this.http.get<IStockRotation[]>(`${this.resourceUrl}/category`, { params, observe: 'response' });
  }

  /**
   * Get stock rotation filtered by ABC classification
   */
  getStockRotationByABCClassification(categorieABC: CategorieABC): Observable<EntityArrayResponseType> {
    const params = new HttpParams().set('categorieABC', categorieABC);
    return this.http.get<IStockRotation[]>(`${this.resourceUrl}/abc`, { params, observe: 'response' });
  }

  /**
   * Get count of products by ABC classification
   */
  getStockRotationCountByABCClassification(): Observable<CountResponseType> {
    return this.http.get<Record<CategorieABC, number>>(`${this.resourceUrl}/count`, { observe: 'response' });
  }

  /**
   * Get slow moving products (Category C)
   */
  getSlowMovingProducts(): Observable<EntityArrayResponseType> {
    return this.http.get<IStockRotation[]>(`${this.resourceUrl}/slow`, { observe: 'response' });
  }

  /**
   * Export stock rotation report as PDF
   */
  exportStockRotationToPdf(): Observable<HttpResponse<Blob>> {
    return this.http.get(`${this.resourceUrl}/export`, { observe: 'response', responseType: 'blob' });
  }
}
